### DISCUSSION TITLE
RandomAffine has weird behavior on Volume dataset

### DISCUSSION THREAD

**Discussion by Khoa-NT** [TorchIO-project/torchio#582] [Q&A] [answered]
Hi, 

I want to do RandomRotation so I follow #229. 
I tried RandomAffine on my Connectome volume dataset (125, 1250, 1250) (D, H, W) as below

```python
transforms = (
        tio.RandomAffine(
        scales=(1,1),
        degrees=1,
        translation=0,
        isotropic=True,
        default_pad_value=0,
        image_interpolation='linear', # linear, nearest
        p=1),
)
transform = tio.Compose(transforms)
```

The result in one section:
![image](https://user-images.githubusercontent.com/12138158/122771732-a739f500-d2e1-11eb-8c10-d067ff9faeff.png)

The result view on ImageJ
![image](https://user-images.githubusercontent.com/12138158/122771613-8f627100-d2e1-11eb-8891-b25d7debc79e.png)

As you can see, the black regions are the missing parts because of the transform but they are not correct.
For example, the location has a red circle and the projection into each slice.
The XY plane still remains the same while it should has missing parts according to XZ and YZ planes.

**Comment by fepegar** [comment, marked as answer]
Hi, @Khoa-NT. Can you please share the image and the code you used to generate that result? 

Those values within the red circle might be artifacts from the 3D rendering algorithm.

**Comment by Khoa-NT** [comment]
Yes, you can download the volume image [here](https://hcmuteduvn-my.sharepoint.com/:u:/g/personal/41201710_hcmut_edu_vn/EcPGKO-xoXJNiPBn5N9-nUsB2HuFjfb76xRzPms9uZ5MFQ?e=dGg40R) 

```python
import numpy as np
import torchio as tio
from skimage.io import imread

img_vol = imread('image_volume.tif')

subject = tio.Subject(
    img_vol=tio.ScalarImage(tensor=np.expand_dims(img_vol, 0).transpose(0,2,3,1)),
)

transform = tio.RandomAffine(
scales=(1,1),
degrees=1,
translation=0,
isotropic=True,
default_pad_value=0,
image_interpolation='linear', # linear, nearest
p=1)

transformed = transform(subject)
```

**Reply by fepegar** [reply]
Thanks for the code. I think everything is working fine and you are just seeing some rendering artifacts. By the way, you don't need `skimage` to load the image. This code works for me:

```python
import torchio as tio

image = tio.ScalarImage('image_volume.tif')
transform = tio.RandomAffine(scales=0, degrees=1)
transformed = transform(image)
transformed.save('image_volume_affine.tif')
```

I used 3D Slicer for visualization.

**Reply by Khoa-NT** [reply]
Thank you for supporting me.

> I used 3D Slicer for visualization.

I tried 3D Slicer but I don't know how to show the 3D render on it. It only shows 3 planes

> By the way, you don't need `skimage` to load the image. This code works for me:

Thank you. I will load the volume directly. 

Can I ask one more question about [RandomFlip](https://torchio.readthedocs.io/transforms/augmentation.html#randomflip)?
I asked in Slack but no one reply yet
What is the difference between `flip_probability` and `p` in RandomFlip?
`p` is in charge of the RandomFlip and `flip_probability` for 3 axes?

**Reply by fepegar** [reply]
> I tried 3D Slicer but I don't know how to show the 3D render on it. It only shows 3 planes

This should get you started: [Volume rendering](https://slicer.readthedocs.io/en/latest/user_guide/modules/volumerendering.html).

> What is the difference between flip_probability and p in RandomFlip?
p is in charge of the RandomFlip and flip_probability for 3 axes?

Yes.

1. A number is chosen with uniform probability between 0 and 1. If it is greater than `p`, the transform won't be executed
2. Inside the transform, a number will be sampled for each axis and compared to `flip_probability` instead.

Does that make sense?

**Reply by Khoa-NT** [reply]
Yes, It's clear now.
Thank you for clarifying.
